### ISSUE TITLE
Meta Flash Attention min OS version?

### ISSUE THREAD

**Issue by brappier** [liuliu/s4nnc#19]
What is the minimum OS version for Metal Flash Attention? 
It does not seem to work with MacOS 12 or 11

**Comment by brappier** [comment]
It gives some shape error.

**Comment by liuliu** [comment]
Minimal is macOS 13.0. If you want attention mask, I believe it is macOS 13.2.

**Comment by liuliu** [comment]
The check we do right now:
```

  public static let isMFACausalAttentionMaskSupported: Bool = {
    #if arch(i386) || arch(x86_64) || !canImport(UIKit)
      return false
    #else
      if #available(iOS 16.5, macOS 13.4, macCatalyst 16.5, *) {
        return true
      }
      return false
    #endif
  }()
  public static let isMFASupported: Bool = {
    #if arch(i386) || arch(x86_64) || !canImport(UIKit)
      return false
    #else
      if #available(iOS 16, macOS 13, macCatalyst 16, *) {
        if let device = MTLCreateSystemDefaultDevice(), device.supportsFamily(.apple7) {
          return true
        }
        return false
      }
      return false
    #endif
  }()
  public static let isMFAGEMMFaster: Bool = {
    #if arch(i386) || arch(x86_64) || !canImport(UIKit)
      return false
    #else
      if #available(iOS 16, macOS 13, macCatalyst 16, *) {
        if let device = MTLCreateSystemDefaultDevice(), device.supportsFamily(.apple7) {
          // MFA GEMM is slower on apple9.
          return device.supportsFamily(.apple9) ? false : true
        }
        return false
      }
      return false
    #endif
  }()
```

**Comment by brappier** [comment]
Thanks
